polygeist/tools/cgeist/Test/Verification/sycl/pure.cpp

```diff
@@ -0,0 +1,20 @@
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -emit-mlir %s -o - | FileCheck %s --check-prefix=CHECK-MLIR
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -S -emit-llvm -fsycl-targets=spir64-unknown-unknown-syclmlir %s -o - | FileCheck %s --check-prefix=CHECK-LLVM
+
+// CHECK-MLIR-DAG: func.func private @func_const() {{.*}} attributes {{{.*}}passthrough = [{{.*}}"nounwind", {{.*}}"willreturn", {{.*}}["memory", "0"]{{.*}}]}
+// CHECK-MLIR-DAG: func.func private @func_pure() {{.*}} attributes {{{.*}}passthrough = [{{.*}}"nounwind", {{.*}}"willreturn", {{.*}}["memory", "21"]{{.*}}]}
+
+// CHECK-LLVM-DAG: declare spir_func i32 @func_const() #[[FUNCATTRS1:[0-9]+]]
+// CHECK-LLVM-DAG: declare spir_func i32 @func_pure() #[[FUNCATTRS2:[0-9]+]]
+
+// CHECK-LLVM-DAG: attributes #[[FUNCATTRS1]] = { {{.*}}nounwind {{.*}}willreturn memory(none){{.*}} }
+// CHECK-LLVM-DAG: attributes #[[FUNCATTRS2]] = { {{.*}}nounwind {{.*}}willreturn memory(read){{.*}} }
+
+#include <sycl/sycl.hpp>
+
+extern "C" SYCL_EXTERNAL int func_const() __attribute__((const));
+extern "C" SYCL_EXTERNAL int func_pure() __attribute__((pure));
+
+extern "C" SYCL_EXTERNAL int foo() {
+  return func_const() + func_pure();
+}
```

Comment:
This is a test not applying to SYCL exclusively (plain C/C++), so I suggest testing with `cgeist` and moving to parent directory. Also, filename is a bit incomplete (maybe `pure_const_attrs`?) Needed changes:
```suggestion
// RUN: cgeist -O0 -w %s --function=* -S | FileCheck %s --check-prefix=CHECK-MLIR
// RUN: cgeist -O0 -w %s --function=* -S -emit-llvm  | FileCheck %s --check-prefix=CHECK-LLVM

// CHECK-MLIR-DAG: func.func private @func_const() {{.*}} attributes {{{.*}}passthrough = [{{.*}}"nounwind", {{.*}}"willreturn", {{.*}}["memory", "0"]{{.*}}]}
// CHECK-MLIR-DAG: func.func private @func_pure() {{.*}} attributes {{{.*}}passthrough = [{{.*}}"nounwind", {{.*}}"willreturn", {{.*}}["memory", "21"]{{.*}}]}

// CHECK-LLVM-DAG: declare spir_func i32 @func_const() #[[FUNCATTRS1:[0-9]+]]
// CHECK-LLVM-DAG: declare spir_func i32 @func_pure() #[[FUNCATTRS2:[0-9]+]]

// CHECK-LLVM-DAG: attributes #[[FUNCATTRS1]] = { {{.*}}nounwind {{.*}}willreturn memory(none){{.*}} }
// CHECK-LLVM-DAG: attributes #[[FUNCATTRS2]] = { {{.*}}nounwind {{.*}}willreturn memory(read){{.*}} }

[[gnu::const]] int func_const();
[[gnu::pure]] int func_pure();

int foo() {
  return func_const() + func_pure();
}
```
I also changed to use standard attribute syntax. Did not check, but this should work as is.

Reply:
In `setMLIRFunctionAttributes()`, we early exited for non sycl device functions, so some attributes are not added for general functions. IIRC, it is to avoid modifying a lot of lit tests.

Reply:
Maybe it's worth it enabling this for the bigger scope in a later PR. LGTM